vcl_templates/_multivariate_tests.vcl.erb

```diff
@@ -4,58 +4,112 @@
+      set var.denominator_<%= test %> += var.nominator_<%= test %>_<%= variant %>;
 <% end -%>
 <% variants.each_with_index do |variant, idx| -%>
 <% if variant == variants.first -%>
-        set var.denominator_<%= test %>_<%= variant %> = var.denominator_<%= test %>;
+      set var.denominator_<%= test %>_<%= variant %> = var.denominator_<%= test %>;
 <% elsif variant != variants.last -%>
 <% previous_variant = variants[idx - 1] -%>
-        set var.denominator_<%= test %>_<%= variant %> = var.denominator_<%= test %>_<%= previous_variant %>;
-        set var.denominator_<%= test %>_<%= variant %> -= var.nominator_<%= test %>_<%= previous_variant %>;
+      set var.denominator_<%= test %>_<%= variant %> = var.denominator_<%= test %>_<%= previous_variant %>;
+      set var.denominator_<%= test %>_<%= variant %> -= var.nominator_<%= test %>_<%= previous_variant %>;
 <% end # if -%>
 <% end # variants.each_with_index -%>
 <% variants.each do |variant| -%>
 <% if variant == variants.first -%>
-        if (randombool(var.nominator_<%= test %>_<%= variant %>, var.denominator_<%= test %>_<%= variant %>)) {
-          set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
+      if (randombool(var.nominator_<%= test %>_<%= variant %>, var.denominator_<%= test %>_<%= variant %>)) {
+        set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
 <% elsif variant != variants.first && variant != variants.last -%>
-        } else if (randombool(var.nominator_<%= test %>_<%= variant %>, var.denominator_<%= test %>_<%= variant %>)) {
-          set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
+      } else if (randombool(var.nominator_<%= test %>_<%= variant %>, var.denominator_<%= test %>_<%= variant %>)) {
+        set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
 <% else -%>
-        } else {
-          set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
-        }
+      } else {
+        set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
+      }
 <% end -%>
 <% end # variants.each-%>
-      }
     }
+<% end # if CookielessAATest -%>
+  }
 <% end # test_config.each -%>
 <% end # ab_tests.each -%>
+}
+else {
+<% ab_tests.each do |test_config| -%>
+<% test_config.each do |test, variants| -%>
+<% if test == "CookielessAATest" -%>
+  if (table.lookup(active_ab_tests, "CookielessAATest") == "true") {
+    if (req.http.User-Agent ~ "^GOV\.UK Crawler Worker") {
+      set req.http.GOVUK-ABTest-<%= test %> = "<%= variants.last %>";
+<% variants.each do |variant| -%>
+    } else if (req.url ~ "[\?\&]ABTest-<%= test %>=<%= variant %>(&|$)") {
+      # Some users, such as remote testers, will be given a URL with a query string
+      # to place them into a specific bucket.
+      set req.http.GOVUK-ABTest-<%= test %> = "<%= variant %>";
+<% end -%>
+    } else {
+      declare local var.denominator_<%= test %> INTEGER;
+      set req.http.Client-Bucket-Full = digest.hash_md5(client.ip req.http.User-Agent);
```

Comment:
> When a request is first handled by a Fastly data center, client.ip represents the IP of the end user. In configurations that use shielding, however, the value of client.ip when the request is processed by the shield data center will be the IP of the edge data center. [source](https://developer.fastly.com/reference/vcl/variables/client-connection/client-ip/)

We don't have shielding enabled, but we might in future, so you may want to use `digest.hash_md5(req.http.Fastly-Client-IP req.http.User-Agent` if this goes beyond a spike.

Reply:
Thanks Bill, I wasn't aware of this 👍